adapters/repos/db/lsmkv/segment_group.go

```diff
@@ -292,22 +292,20 @@ func (sg *SegmentGroup) count() int {
 }
 
 func (sg *SegmentGroup) shutdown(ctx context.Context) error {
-	sg.maintenanceLock.Lock()
-	defer sg.maintenanceLock.Unlock()
-
-	compactionHalted := make(chan struct{})
-
-	go func() {
-		sg.compactionCycle.Stop(ctx)
-		compactionHalted <- struct{}{}
-	}()
-
-	select {
-	case <-ctx.Done():
+	if err := sg.compactionCycle.StopAndWait(ctx); err != nil {
 		return errors.Wrap(ctx.Err(), "long-running compaction in progress")
-	case <-compactionHalted:
 	}
 
+	// Lock acquirement placed after compaction cycle stop request, due to occasional deadlock,
+	// because compaction logic used in cycle also requires maintenance lock.
```

Comment:
Great example of a comment outlining the motivation behind why a change was made. ❤️